crates/sui-framework/packages/sui-framework/sources/metadata_value.move

```diff
@@ -0,0 +1,266 @@
+
+/// Creates a new `MetadataValue` representing a `String` value.
+public fun new_string(value: String): MetadataValue { MetadataValue::String(value) }
+
+/// Creates a new `MetadataValue` representing an `ID` value.
+public fun new_id(value: ID): MetadataValue { MetadataValue::ID(value) }
+
+/// Creates a new `MetadataValue` representing a generic `BCS` value.
+/// Should only be used for values that are not already represented by the other
+/// constructors, add challenge in reading the value as well as indexing.
+public fun new_bcs<T: copy + drop + store>(value: T): MetadataValue {
+    MetadataValue::BCS(type_name::get<T>(), bcs::to_bytes(&value))
+}
+
+// === Is Variant ===
+
+/// Checks if the `MetadataValue` is a `bool` value.
+public fun is_bool(value: &MetadataValue): bool {
+    match (value) {
+        MetadataValue::Bool(_) => true,
+        _ => false,
+    }
+}
+
+/// Checks if the `MetadataValue` is a `u8` value.
+public fun is_u8(value: &MetadataValue): bool {
+    match (value) {
+        MetadataValue::U8(_) => true,
+        _ => false,
+    }
+}
+
+/// Checks if the `MetadataValue` is a `u16` value.
+public fun is_u16(value: &MetadataValue): bool {
+    match (value) {
+        MetadataValue::U16(_) => true,
+        _ => false,
+    }
+}
+
+/// Checks if the `MetadataValue` is a `u32` value.
+public fun is_u32(value: &MetadataValue): bool {
+    match (value) {
+        MetadataValue::U32(_) => true,
+        _ => false,
+    }
+}
+
+/// Checks if the `MetadataValue` is a `u64` value.
+public fun is_u64(value: &MetadataValue): bool {
+    match (value) {
+        MetadataValue::U64(_) => true,
+        _ => false,
+    }
+}
+
+/// Checks if the `MetadataValue` is a `u128` value.
+public fun is_u128(value: &MetadataValue): bool {
+    match (value) {
+        MetadataValue::U128(_) => true,
+        _ => false,
+    }
+}
+
+/// Checks if the `MetadataValue` is a `u256` value.
+public fun is_u256(value: &MetadataValue): bool {
+    match (value) {
+        MetadataValue::U256(_) => true,
+        _ => false,
+    }
+}
+
+/// Checks if the `MetadataValue` is an `address` value.
+public fun is_address(value: &MetadataValue): bool {
+    match (value) {
+        MetadataValue::Address(_) => true,
+        _ => false,
+    }
+}
+
+/// Checks if the `MetadataValue` is a `String` value.
+public fun is_string(value: &MetadataValue): bool {
+    match (value) {
+        MetadataValue::String(_) => true,
+        _ => false,
+    }
+}
+
+/// Checks if the `MetadataValue` is an `ID` value.
+public fun is_id(value: &MetadataValue): bool {
+    match (value) {
+        MetadataValue::ID(_) => true,
+        _ => false,
+    }
+}
+
+/// Checks if the `MetadataValue` is a generic `BCS` value.
+public fun is_bcs(value: &MetadataValue): bool {
+    match (value) {
+        MetadataValue::BCS(_, _) => true,
+        _ => false,
+    }
+}
+
+// === Getters ===
```

Comment:
Do we also want `try_read_<ty>` operations too? OR do you think it's not needed in the presence of the predicates? Or possibly should we have these instead of the predicates, and then the predicates are just `x.try_get_<ty>.is_none()`?

Reply:
We could, that's an option. But I thought the whole idea was that you must know a type for a key, no?

Reply:
And it will be missing for `read_bcs`.